plugin/db/mysql/restore.go

```diff
@@ -172,9 +173,33 @@ func (driver *Driver) replayBinlog(ctx context.Context, originalDatabase, pitrDa
 
 	mysqlbinlogCmd.Stderr = os.Stderr
 
+	mysqlStdin := common.NewCountingReader(mysqlRead)
 	mysqlCmd.Stderr = os.Stderr
 	mysqlCmd.Stdout = os.Stderr
-	mysqlCmd.Stdin = mysqlRead
+	mysqlCmd.Stdin = mysqlStdin
+
+	// Update replay binlog progress
+	totalBinlogBytes, err := getFileSizeSum(replayBinlogPaths)
+	if err != nil {
+		return fmt.Errorf("failed to get file size sum of replay binlog files, error: %w", err)
+	}
+	driver.replayBinlogProgress = 0
+	stopChan := make(chan struct{})
+	go func() {
+		ticker := time.NewTicker(1 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				progress := mysqlStdin.Count() * 10000 / totalBinlogBytes
+				atomic.StoreUint64(&driver.replayBinlogProgress, progress)
+			case <-ctx.Done():
+				return
+			case <-stopChan:
+				return
+			}
+		}
+	}()
```

Comment:
What would happen to this go routine if `replayBinlog()` encounters an error and returns at line 205? I doubt that the go routine will never get a chance to return.

Reply:
Good catch.
`defer func() { close(stopChan )} ()` seems to solve this problem.

Reply:
`defer close(stopChan)`

Reply:
Oops...